lib/puppet/pops/loader/task_instantiator.rb

```diff
@@ -1,72 +1,25 @@
-    else
-      json_text = loader.get_contents(metadata)
-      begin
-        Puppet::Util::Json.load(json_text).freeze || EMPTY_HASH
-      rescue Puppet::Util::Json::ParseError => ex
-        raise Puppet::ParseError.new(ex.message, metadata)
-      end
-    end
-  end
-
-  def self.validate_implementations(typed_name, directory, metadata, executables)
-    name = typed_name.name
-    basename = typed_name.name_parts[1] || 'init'
-    # If 'implementations' is defined, it needs to mention at least one
-    # implementation, and everything it mentions must exist.
-    if metadata.key?('implementations')
-      if metadata['implementations'].is_a?(Array)
-        metadata['implementations'].map do |impl|
-          path = executables.find { |real_impl| File.basename(real_impl) == impl['name'] }
-          if path
-            { "name" => impl['name'], "requirements" => impl.fetch('requirements', []), "path" => path }
-          else
-            raise ArgumentError, _("Task metadata for task %{name} specifies missing implementation %{implementation}") %
-              { name: name, implementation: impl['name'] }
-          end
-        end
-      else
-        # If 'implementations' is the wrong type, we just pass it through and
-        # let the task type definition reject it.
-        metadata['implementations']
-      end
-    # If implementations isn't defined, then we use executables matching the
-    # task name, and only one may exist.
-    else
-      implementations = executables.select { |impl| File.basename(impl, '.*') == basename }
-      if implementations.empty?
-        raise ArgumentError, _('No source besides task metadata was found in directory %{directory} for task %{name}') %
-          { name: name, directory: directory }
-      elsif implementations.length > 1
-        raise ArgumentError, _("Multiple executables were found in directory %{directory} for task %{name}; define 'implementations' in metadata to differentiate between them") %
-          { name: name, directory: implementations[0] }
-      end
-
-      [{ "name" => File.basename(implementations.first), "path" => implementations.first, "requirements" => [] }]
-    end
-  end
-
   def self.create(loader, typed_name, source_refs)
     name = typed_name.name
     basename = typed_name.name_parts[1] || 'init'
     dirname = File.dirname(source_refs[0])
     metadata_files, executables = source_refs.partition { |source_ref| source_ref.end_with?('.json') }
     metadata_file = metadata_files.find { |source_ref| File.basename(source_ref, '.json') == basename }
 
-    metadata = load_metadata(loader, metadata_file)
+    metadata = Puppet::Module::Task.read_metadata(metadata_file)
 
-    implementation_metadata = validate_implementations(typed_name, dirname, metadata, executables)
+    implementations = Puppet::Module::Task.find_implementations(name, dirname, metadata, executables)
+    files = Puppet::Module::Task.find_files(metadata)
 
     arguments = {
```

Comment:
This won't work, because we mask the implementations and files entries from metadata, and `Puppet::Module::Task` expects metadata to be separate from file info.